### ISSUE TITLE
Revisit property syncing and API

### ISSUE THREAD

**Issue by almarklein** [zoofIO/flexx#123]
I am not entirely happy with the API for determining where props are validated / synced. I find it results in some confusion. Writing down some thoughts here. Feel free to comment. 

A prop can be:

* A: Python only - stuff that can not be serialized or the client should be unaware of.
* B: Validated in Python, proxied in JS - useful for stuff that only the server can do, but prevents exporting to a standalone HTML document. 
* C: JS only - useful for stuff that changes a lot, to avoid traffic.
* D: Validated in JS, proxied in Py - useful when it can only be validated in JS (e.g. Bokeh plot).
* E: Validated in both Python and JS - useful because is immediately set and can still be used as export HTML.

Observations:

* I find that E is the most common case, but need to write `@prop(both=True)` which is verbose and a bit confusing. 
* A and C are currently achieved through `@prop(sync=False)`.
* Need a way to achieve C with existing properties (e.g. `x` in a fastly updating plotwidget), see #112.
* B and D are kind of the default, right now, even though E is often preferred.

Earlier, I tried this:
```python
class Foo(app.Model):
     # both-stuff here
     class Py:
        # Python specific here
     class JS:
       # Js specifics here
```
But this was impossible to achieve (or at least hard enough to make a meta-class hell).

Maybe this?
```python
class Foo(app.Model):
     # Python stuff here
     class Both:
        # Both-props here
     class JS:
       # Js specifics here
```

**Comment by Korijn** [comment]
How about more explicit flags with more familiar terminology on the property decorator? E.g.:

```python
@prop(read='py,js', write='py,js') # case E
```

**Comment by almarklein** [comment]
> What makes the property decorator confusing to me is that I suddenly need to think about the underlying mechanisms of flexx.

Well said. I think that is the main problem.

> can't you work out from the dependency graph (of events and handlers) what settings should be applied to properties

For events that's done, yes (events only get send to the other side if there are registered handlers). But for properties, you cannot know (unless you analyze the code) what properties are used where.

> Although automatic connection of foo to on_foo caught me off guard as well :)

See #124

**Comment by almarklein** [comment]
Another idea (going a bit into brainstorm mode): Allow only methods and class attributes in the JS class, and force *all* props to be defined on the main class. Let props have a `where` flag, that is "both" by default, and can also be set to "python" (A/B) or "js" (C/D).

**Comment by almarklein** [comment]
After discussion with @Korijn this seems to make things a lot simpler:

```python
class SomeWidget(ui.Widget):
    
    def init(self):
        """ Construct compound widget here. This method is only run in
        Python, but the sub-widgets will also exist on the JS end.
        """
        with ui.HBox():
            self.but = ui.Button()
            self.img = ui.ImageWidget()
    
    @event.prop
    def foo(self, v=''):
        """ Exists and is validated on both ends """
        return str(v)
    
    @event.prop
    def bar(self, v):
        """ Exists on both ends, but is validated on one end """
        if this_is_js():
            return v  # or maybe raise DontSetThisProperty ?
        else:
            return validate(v)
    
    class Py:
        
        def init(self):
            """ Initialize server specifics here """
        
        @event.prop
        def bert(self, v):
            """ Only exists in Py """
            return str(v)
        
        def define_py_only_methods_and_handlers_here(self):
            pass  # if the Py class has no handlers, the model is likely exportable
    
    class JS:
        
        def init(self):
            """ Initialize client specifics here """
        
        @event.prop
        def ernie(self, v):
            """ Only exists in JS """
            return str(v)

        def define_js_only_methods_and_handlers_here(self):
            pass
```

* In the root of the class, there is only stuff that applies to both Python and JS. Python and JS specifics are implemented in the corresponding "sub classes".
* The `init()` in the root is intended to create sub-widgets. We cannot prevent a user from doing server-specific, but that sort of thing *should* happen in `Py.init()`. 
* There are possibly 3 instances of `init()`, which *might* be confusing ("does that not cause a name clash?"), we'll see how that works out, if needed we can rename the py/js specific `init()` to `post_init()` or something.
* There is no more need for *any* property flags :D
* It is now possible to define props that only exist on one end, which is similar to the behavior provided with the `sync` flag, but better. A nice thing is that one can still connect to the corresponding event.
* By making use of `this_is_js()` in the validator, its easy to make "proxy" properties. Maybe push this a bit further by allow raising a specific error in the validator so that the prop value is not actually set (until it round-trips).

**Comment by almarklein** [comment]
Arg. Now I remember the problem with having a `Foo.Py` class: 1) you want to expose `Foo.Py` as `Foo` in Python. This is resolved in the above proposal by simply copying class attributes over. But there is also 2) calling `super()` in a method in `Foo.Py` will use `Foo.Py` as its class, not `Foo`. IIRC fixing this with the metaclass seemed rather a rabbithole.

Forcing users to use `super(Foo, self)` would work, but that feels too much like legacy Py to my taste ;)

I will have a brief look (again) to see if there is a workaround. Otherwise I think we should do:

```
class Foo(app.Model):
    def init(self):
         ...
    
    @event.prop
    def bert(self, v):
        """ Only exists in Py """
        return str(v)

    class Both:
        @event.prop
        def foo(self, v=''):
            """ Exists and is validated on both ends """
            return str(v)

        @event.prop
        def bar(self, v):
            """ Exists on both ends, but is validated on one end """
            if this_is_js():
                return v  # or maybe raise DontSetThisProperty ?
            else:
                return validate(v)

    class JS:
        @event.prop
         def ernie(self, v):
          """ Only exists in JS """
              return str(v)
```

This has the same benefits as the previous proposal (e.g. no more property flags), but changes where py-only and common class attributes are defined. This would also resolve the potential confusion with the `init()`. BTW, this still suffers that `super()` cant be used inside the `Both` subclass, but in general the `Both` subclass will only contain properties, so that should not be a problem.

**Comment by Korijn** [comment]
Can't you just disallow init() and super() in Foo.Py?

**Comment by almarklein** [comment]
Disallowing `super()` would be tricky, but more importantly, I don't think we want to prevent users from calling a superclass' method in Python (and `Model` users super).

**Comment by almarklein** [comment]
Closed by #130
